### PR TITLE
Issue 666: Introduce registration client for bookkeeper client to do bookie discover

### PR DESCRIPTION
**Pull request by zhaijack**
Descriptions of the changes in this PR:
while in BOOKKEEPER-628, It is trying to improve/generalize the bookie registration process.

This PR follows the work of #663, which provide bookie side registration, and it is for client side work, it tries to introduce RegistrationClient for bookkeeper client to do bookie discover.

This PR follows the work of #663, which provide bookie side registration, 

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
> 
> - [x] Make sure the PR title is formatted like:
>     `<Issue # or BOOKKEEPER-#>: Description of pull request`
>     `e.g. Issue 123: Description ...`
>     `e.g. BOOKKEEPER-1234: Description ...`
> - [x] Make sure tests pass via `mvn clean apache-rat:check install findbugs:check`.
> - [x] Replace `<Issue # or BOOKKEEPER-#>` in the title with the actual Issue/JIRA number.
> 
> ---

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/bookkeeper-precommit-pullrequest-docker/153/

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java [hunk 1]
```diff
@@ -446,6 +446,7 @@ public void handleBookiesThatJoined(Set<BookieSocketAddress> joinedBookies) {
             Ensemble<BookieNode> parentEnsemble,
             Predicate<BookieNode> parentPredicate) throws BKNotEnoughBookiesException {
         rwLock.readLock().lock();
+
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java]: seems unused blank lines?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 2]
```diff
@@ -99,6 +99,8 @@
     // Kluge for compatibility testing. Never set this outside tests.
     public final static String LEDGER_MANAGER_FACTORY_DISABLE_CLASS_CHECK = "ledgerManagerFactoryDisableClassCheck";
 
+
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: seems unused blank lines?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 3]
```diff
@@ -485,4 +487,6 @@ public AbstractConfiguration setTLSEnabledProtocols(
     public String getTLSEnabledProtocols() {
         return getString(TLS_ENABLED_PROTOCOLS, null);
     }
+
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: seems unused blank lines?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/http/BKHttpServiceProvider.java [hunk 4]
```diff
@@ -59,7 +59,7 @@
     private BKHttpServiceProvider(BookieServer bookieServer,
                                   AutoRecoveryMain autoRecovery,
                                   ServerConfiguration serverConf)
-        throws IOException, KeeperException, InterruptedException {
+        throws IOException, KeeperException, InterruptedException, BKException {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/http/BKHttpServiceProvider.java]: can we remove KeeperException here?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieServer.java [hunk 5]
```diff
@@ -101,7 +102,7 @@ protected Bookie newBookie(ServerConfiguration conf)
                 new Bookie(conf, statsLogger.scope(BOOKIE_SCOPE));
     }
 
-    public void start() throws IOException, UnavailableException, InterruptedException, KeeperException {
+    public void start() throws IOException, UnavailableException, InterruptedException, KeeperException, BKException {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieServer.java]: can we remove `KeeperException` here?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java [hunk 6]
```diff
@@ -111,21 +111,21 @@ private int entryCount(long ledgerId, ServerConfiguration bookieConf,
      */
     @Test
     public void testSingleMessageAuth() throws Exception {
-        ServerConfiguration bookieConf = newServerConfiguration();
-        bookieConf.setBookieAuthProviderFactoryClass(
+            ServerConfiguration bookieConf = newServerConfiguration();
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java]: wrong indent?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieInitializationTest.java [hunk 7]
```diff
@@ -89,6 +89,7 @@ public BookieInitializationTest() {
     @Override
     public void setUp() throws Exception {
         super.setUp();
+        LOG.info("setUp");
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieInitializationTest.java]: remove this?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestReadTimeout.java [hunk 8]
```diff
@@ -52,38 +52,39 @@ public TestReadTimeout() {
     @SuppressWarnings("deprecation")
     @Test
     public void testReadTimeout() throws Exception {
-        final AtomicBoolean completed = new AtomicBoolean(false);
+            final AtomicBoolean completed = new AtomicBoolean(false);
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestReadTimeout.java]: wrong indent?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java (hunk 1 above)
Comment (zhaijack) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java]: Thanks. will remove it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java (hunk 2 above)
Comment (zhaijack) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: Thanks. will remove it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java (hunk 3 above)
Comment (zhaijack) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: Thanks. will remove it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/http/BKHttpServiceProvider.java (hunk 4 above)
Comment (zhaijack) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/http/BKHttpServiceProvider.java]: Thanks. will try to remove it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieServer.java (hunk 5 above)
Comment (zhaijack) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieServer.java]: Thanks. will try to remove it.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieInitializationTest.java (hunk 7 above)
Comment (zhaijack) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieInitializationTest.java]: Thanks. will  remove it.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java (hunk 6 above)
Comment (zhaijack) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java]: Thanks. will fix it.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestReadTimeout.java (hunk 8 above)
Comment (zhaijack) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestReadTimeout.java]: Thanks. will fix it.

**Comment by sijie** [comment]
retest it please

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
retest this please

**Pull request closed without merging**